test/Microsoft.AspNet.Mvc.FunctionalTests/TagHelpersTest.cs

```diff
@@ -148,5 +148,110 @@ public class TagHelpersTests
+        {
+            // Arrange
+            var server = TestServer.Create(_provider, _app);
+            var client = server.CreateClient();
+            client.BaseAddress = new Uri("http://localhost");
+            client.DefaultRequestHeaders.Add("Locale", "N");
+
+            // Act - 1
+            // Verify that content gets cached based on vary-by-params
+            var response1 = await client.GetStringAsync("/catalog?categoryId=1&correlationid=1");
+
+            // Assert - 1
+            var expected1 =
+@"<h2>Category: Laptops</h2>
+<h2>Region: North</h2>
+
+    <h2>Cached content</h2>
+    Locations closest to your locale:
+
+NorthWest Store
+<div>CorrelationId in View Component: 1</div>
+
+    <partial-title>Listing items</partial-title>
+
+Cached Content for Laptops
+<div>CorrelationId in Partial: 1</div>
+
+    <div>CorrelationId in Splash: 1</div>";
+            Assert.Equal(expected1, response1.Trim());
+
+            // Act - 2
+            // Verify content gets changed in partials when one of the vary by parameters is changed
+            var response2 = await client.GetStringAsync("/catalog?categoryId=3&correlationid=2");
+
+            // Assert - 2
+            var expected2 =
+@"<h2>Category: Phones</h2>
+<h2>Region: North</h2>
+
+    <h2>Cached content</h2>
+    Locations closest to your locale:
+
+NorthWest Store
+<div>CorrelationId in View Component: 1</div>
+
+    <partial-title>Listing items</partial-title>
+
+Cached Content for Phones
+<div>CorrelationId in Partial: 2</div>
+
+    <div>CorrelationId in Splash: 2</div>";
+
+            Assert.Equal(expected2, response2.Trim());
+
+            // Act - 3
+            // Verify content gets changed in a View Component when the Vary-by-header parameters is changed
+            client.DefaultRequestHeaders.Add("Locale", "East");
+            var response3 = await client.GetStringAsync("/catalog?categoryId=3&correlationid=3");
+
+            // Assert - 2
+            var expected3 =
+@"<h2>Category: Phones</h2>
+<h2>Region: Default</h2>
+
+    <h2>Cached content</h2>
+    Locations closest to your locale:
+
+Nationwide Store
+<div>CorrelationId in View Component: 3</div>
+
+    <partial-title>Listing items</partial-title>
+
+Cached Content for Phones
+<div>CorrelationId in Partial: 2</div>
+
+    <div>CorrelationId in Splash: 3</div>";
+            Assert.Equal(expected3, response3.Trim());
+        }
+
+        [Fact]
+        public async Task CacheTagHelper_ExpiresContent_BasedOnExpiresParameter()
+        {
+            // Arrange
+            var server = TestServer.Create(_provider, _app);
+            var client = server.CreateClient();
+            client.BaseAddress = new Uri("http://localhost");
+
+            // Act - 1
+            var response1 = await client.GetStringAsync("/catalog/2");
+
+            // Assert - 1
+            var expected1 = "Cached content for 2";
+            Assert.Equal(expected1, response1.Trim());
+
+            // Act - 2
```

Comment:
suggest adding an immediate Act & Assert 1a which confirms the same response comes back then